### PR TITLE
feat (ui): Add Native Histogram rendering to new UI

### PR DESCRIPTION
**Pull request by Maniktherana**
Adds native histogram to new ui
Some changes:
 - Bucket table underneath the histogram is scrollable
 - Now hovering the entire DataTable.tsx component doesnt highlight it

### CODE REVIEW

**Comment by Maniktherana** [comment]
https://github.com/prometheus/prometheus/assets/14011425/a8b7b8db-b3ca-4fcb-b40e-9dcb77bb683e

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx [hunk 1]
```diff
@@ -111,7 +127,34 @@ const DataTable: FC<DataTableProps> = ({ expr, evalTime, retriggerIdx }) => {
...
+                          <span>
+                            <strong>Total count:</strong> {s.histogram[1].count}
+                          </span>
+                          <span>
+                            <strong>Sum:</strong> {s.histogram[1].sum}
+                          </span>
+                        </div>
+                        <div className={classes.histogramSummary}>
+                          <span>x-axis scale:</span>
+                          <SegmentedControl
```
Comment (juliusv) [inline web/ui/mantine-ui/src/pages/query/DataTable.tsx:148]: Let's add `size="xs"` to this one to fit the sizes around it better.

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx [hunk 2]
```diff
@@ -161,4 +204,44 @@ const DataTable: FC<DataTableProps> = ({ expr, evalTime, retriggerIdx }) => {
   );
 };
 
+export const histogramTable = (h: Histogram): ReactNode => (
+  <Table>
+    <Table.Thead>
+      <Table.Tr>
+        <Table.Th style={{ textAlign: "center" }} colSpan={2}>
+          Histogram Sample
```
Comment (juliusv) [inline web/ui/mantine-ui/src/pages/query/DataTable.tsx:212]: The best title for this table is maybe "Bucket counts" now, since you have the total sum + count above the table. But probably it's even ok / better to just omit the title, hopefully the table should be clear.

File: web/ui/mantine-ui/src/pages/query/HistogramChart.module.css [hunk 3]
```diff
@@ -0,0 +1,106 @@
+.histogramYWrapper {
+    display: flex;
```
Comment (juliusv) [inline web/ui/mantine-ui/src/pages/query/HistogramChart.module.css:2]: Same about 2-space vs. 4-space indentation.

File: web/ui/mantine-ui/src/pages/query/HistogramChart.tsx [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+import classes from "./HistogramChart.module.css";
+import { Tooltip } from "@mantine/core";
+
+interface HistogramChartProps {
+  histogram: Histogram;
+  index: number;
+  scale: string;
+}
+
+const HistogramChart: FC<HistogramChartProps> = ({
```
Comment (juliusv) [inline web/ui/mantine-ui/src/pages/query/HistogramChart.tsx:21]: I still noticed one display glitch when switching to linear mode, with a vertical grid line appearing left of my Y axis:

image

File: web/ui/package-lock.json [hunk 5]
```diff
@@ -106,7 +106,7 @@
     },
     "mantine-ui": {
       "name": "@prometheus-io/mantine-ui",
-      "version": "0.51.2",
+      "version": "0.49.1",
```
Comment (juliusv) [inline web/ui/package-lock.json:109]: This change probably shouldn't be part of the PR?

**Review by juliusv — COMMENTED** [review]
Thanks! Looks great overall, just a few comments :)

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx [hunk 6]
```diff
@@ -34,6 +47,9 @@ export interface DataTableProps {
 }
 
 const DataTable: FC<DataTableProps> = ({ expr, evalTime, retriggerIdx }) => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
[the remark this answers is not in the document]
Comment (juliusv) [reply web/ui/mantine-ui/src/pages/query/DataTable.tsx:50]: FWIW the CI checks are also failing because of `error    Unused eslint-disable directive (no problems were reported from '@typescript-eslint/no-unused-vars')`

File: web/ui/mantine-ui/src/pages/query/DataTable.module.css [hunk 7]
```diff
@@ -1,10 +1,23 @@
 .tableWrapper {
-  border: 1px solid
-    light-dark(var(--mantine-color-gray-3), var(--mantine-color-dark-5));
-  border-radius: var(--mantine-radius-default);
+    border: 1px solid
```
[the remark this answers is not in the document]
Comment (Maniktherana) [reply web/ui/mantine-ui/src/pages/query/DataTable.module.css:2]: Hmm, might have something to do with eslint rules?
I suggest we copy the ones in the current UI, they work really well and I've never had issues

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Maniktherana) [reply web/ui/mantine-ui/src/pages/query/DataTable.tsx:50]: Ah yes, while I was working on some stuff I was getting build failures because of unused vars. Forgot to remove this

(even more reason to use older eslint rules)

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx [hunk 8]
```diff
@@ -101,7 +117,7 @@ const DataTable: FC<DataTableProps> = ({ expr, evalTime, retriggerIdx }) => {
         }}
         styles={{ loader: { width: "100%", height: "100%" } }}
       />
-      <Table highlightOnHover fz="xs">
+      <Table fz="xs">
```
[the remark this answers is not in the document]
Comment (Maniktherana) [reply web/ui/mantine-ui/src/pages/query/DataTable.tsx:120]: Perhaps we can do some sort of conditional rendering. I was trying to get it to only highlight the left pane if histograms were rendered but there's no such override available.

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx [hunk 9]
```diff
@@ -111,7 +127,34 @@ const DataTable: FC<DataTableProps> = ({ expr, evalTime, retriggerIdx }) => {
...
                   {s.value && s.value[1]}
-                  {s.histogram && "TODO HISTOGRAM DISPLAY"}
+                  {s.histogram && (
+                    <>
+                      <HistogramChart
+                        histogram={s.histogram[1]}
+                        index={idx}
+                        scale={scale}
+                      />
+                      <div className={classes.histogramSummaryWrapper}>
```
[the remark this answers is not in the document]
Comment (Maniktherana) [reply web/ui/mantine-ui/src/pages/query/DataTable.tsx:137]: Noted, I'll look into mantine docs and see all I can use instead of classes

File: web/ui/mantine-ui/src/pages/query/HistogramChart.tsx (hunk 4 above)
Comment (Maniktherana) [reply web/ui/mantine-ui/src/pages/query/HistogramChart.tsx:21]: Ah I might know whats causing that

I'll try to fix

File: web/ui/package-lock.json (hunk 5 above)
Comment (Maniktherana) [reply web/ui/package-lock.json:109]: This is automatic I believe
Not sure why its happening but I don't really touch lockfiles

File: web/ui/mantine-ui/src/pages/query/DataTable.module.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (juliusv) [reply web/ui/mantine-ui/src/pages/query/DataTable.module.css:2]: > Hmm, might have something to do with eslint rules?

Honestly, I'm not sure - we're doing 2 spaces in both the old and new code bases, and my editor is correctly doing 2 spaces in both cases without a specific rule for it (as far as I'm aware). But that may just be because it's the VSCode default, and adding an explicit rule for it would be even better.

> I suggest we copy the ones in the current UI, they work really well and I've never had issues

If you're talking about https://github.com/prometheus/prometheus/blob/mantine-ui/web/ui/react-app/.eslintrc.json vs. https://github.com/prometheus/prometheus/blob/mantine-ui/web/ui/mantine-ui/.eslintrc.cjs, I'd be happy to copy over specific rules that we like, but a bit wary copying over old cruft that we don't know if we still want.

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (juliusv) [reply web/ui/mantine-ui/src/pages/query/DataTable.tsx:120]: Yeah, I also tried removing it selectively for rows that have native histograms with `mod={{ hover: true }}` (see https://mantine.dev/styles/data-attributes/#mod-prop), but it didn't override the `data-hover="true"` to `false`.

Let's keep it for later :)

File: web/ui/mantine-ui/src/pages/query/DataTable.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (juliusv) [reply web/ui/mantine-ui/src/pages/query/DataTable.tsx:137]: Oh yeah, no need to necessarily change it right now, but would be cool to converge on that more over time I guess.

File: web/ui/package-lock.json (hunk 5 above)
Comment (juliusv) [reply web/ui/package-lock.json:109]: Yep, it's from the build process and depends on the underlying Prometheus version it sees, but no need to check it in unless we cut UI releases.

File: web/ui/mantine-ui/src/pages/query/HistogramChart.tsx (hunk 4 above)
Comment (Maniktherana) [reply web/ui/mantine-ui/src/pages/query/HistogramChart.tsx:21]: Hmm I actually can't seem to recreate this but my guess is that its the the y-axis marker for 0

This might be a stretch: but if you use the server from `main` with latest changes and run the new ui detached, does that help remove it?

**Comment by Maniktherana** [comment]
I've applied all the suggested changes except the one with the floating weird line

File: web/ui/mantine-ui/src/pages/query/HistogramChart.tsx (hunk 4 above)
Comment (juliusv) [reply web/ui/mantine-ui/src/pages/query/HistogramChart.tsx:21]: Huh strange, at the moment I can't reproduce it either anymore, no matter if I use the server from `main` or your PR branch. Maybe it was super data-dependent, but since it happened for all the sub-histograms, that would be weird too.

I guess I'll just keep an eye on it in the future in case it comes back.

**Comment by juliusv** [comment]
Nice, I think this is mostly ready to merge for now, we just have to deal with this CI lint check failing:

```
/__w/prometheus/prometheus/web/ui/mantine-ui/src/pages/query/DataTable.tsx
  205:14  warning  Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components  react-refresh/only-export-components
ESLint found too many warnings (maximum: 0).
```

Maybe just remove the `export`, since the `histogramTable` function isn't used outside of that file anyway?

**Comment by Maniktherana** [comment]
> Maybe just remove the export, since the histogramTable function isn't used outside of that file anyway?

done!

**Comment by juliusv** [comment]
Yay, thanks! :) :+1:
